lib/teiserver/matchmaking/algos.ex

```diff
@@ -1,35 +1,42 @@
 defmodule Teiserver.Matchmaking.Algos do
   @moduledoc """
-  Contains the algorithm for matchmaking.
-  Currently only has the `ignore_os` function, but is going to have
-  at least another one taking OS (and perhaps other parameters) into
-  account
+  Interface for matchmaking algorithms
   """
 
   alias Teiserver.Matchmaking.Member
   alias Teiserver.Helpers.Combi
 
+  @type state :: term()
+
   @doc """
-  The simplest possible algorithm for matchmaking.
-  It only looks at filling the teams with the correct number of players
-  without looking at anything else like wait time or OS.
-  This is useful for testing.
+  A way to initialize the module with some persistent state.
+  For example, a http client or getting some parameters from the DB
+  """
+  @callback init(team_size :: pos_integer(), team_count :: pos_integer()) :: state()
 
-  This returns a list of potential matches.
-  A match is a list of teams, a team is a list of member
+  @doc """
+  The function to invoke to pair some members.
+  It returns a list of valid matches. A match is a list of teams, a team is a
+  list of member.
+
+  Optionally can pass a predicate to filter some matches
   """
-  @spec ignore_os(team_size :: pos_integer(), team_count :: pos_integer(), members: [Member.t()]) ::
-          :no_match | {:match, [[[Member.t()]]]}
-  def ignore_os(team_size, team_count, members) do
-    case get_matches(team_size, team_count, members, []) do
-      [] -> :no_match
-      matches -> {:match, matches}
-    end
-  end
+  @callback get_matches(
+              members :: [Member.t()],
+              state :: state()
+            ) :: :no_match | {:match, [[[Member.t()]]]}
 
-  def get_matches(team_size, team_count, members, acc) do
+  @spec match_members(
+          [Member.t()],
+          team_size :: pos_integer(),
+          team_count :: pos_integer(),
+          pred :: ([[Member.t()]] -> boolean),
+          acc :: term()
+        ) :: [[[Member.t()]]]
+  def match_members(members, team_size, team_count, pred, acc \\ []) do
     res =
-      match_stream(team_size, team_count, members)
+      match_stream(members, team_size, team_count)
+      |> Stream.filter(pred)
```

Comment:
the `take(1)` could be improved a bit as well later. We could `take` more and then get the best according to some metric, but that's future improvements.